subt_ign/src/VisibilityTable.cc

```diff
@@ -348,29 +378,109 @@ bool VisibilityTable::PopulateVisibilityInfoHelper(
       // select the one with lowest cost.
-      bestRouteCost = std::min(bestRouteCost, currentRouteCost);
+      if ((currentRouteCost < bestRouteCost) ||
+          (currentRouteCost == bestRouteCost &&
+           relayToDstNumRelaysCrossed < bestRouteNumRelaysCrossed))
+      {
+        bestRouteCost = currentRouteCost;
+        bestRelay = i;
+        bestRouteNumRelaysCrossed = relayToDstNumRelaysCrossed;
+      }
     }
   }
 
   // Among all posible routes, select the one with lowest cost.
-  _visibilityInfoWithRelays[std::make_pair(_from, _to)] =
-    std::min(srcToDstCost, bestRouteCost);
+  if (srcToDstCost <= bestRouteCost)
+  {
+    _visibilityInfoWithRelays[std::make_pair(_from, _to)] =
+      {srcToDstCost, {}, {}, {}, 0};
+  }
+  else
+  {
+    std::vector<ignition::math::graph::VertexId> path;
+    ignition::math::graph::VertexId relay = bestRelay;
+    double maxDistance = 0;
+    ignition::math::Vector3d firstBreadcrumbPos =
+      ignition::math::Vector3d::Zero;
+    ignition::math::Vector3d lastBreadcrumbPos = ignition::math::Vector3d::Zero;
+
+    if (relay != math::graph::kNullId)
+    {
+      path.push_back(relay);
+      auto relayToDstIt = _visibilityInfoWithRelays.find(
+        std::make_pair(relay, _to));
+      if (relayToDstIt != _visibilityInfoWithRelays.end())
+      {
+        auto otherRelays = relayToDstIt->second.route;
+        path.insert(path.end(), otherRelays.begin(), otherRelays.end());
+
+        // Calculate the greatest distance single hop between breadcrumbs only.
+        maxDistance = this->MaxDistanceSingleHop(
+          _visibilityInfoWithRelays, path, _to);
+      }
+
+      // Update the first and last breadcrumb positions.
+      firstBreadcrumbPos = this->breadcrumbs.at(path.front()).front();
+      lastBreadcrumbPos = this->breadcrumbs.at(path.back()).front();
+    }
+
+    _visibilityInfoWithRelays[std::make_pair(_from, _to)] =
+      {bestRouteCost, path, firstBreadcrumbPos, lastBreadcrumbPos, maxDistance};
+  }
+
+  // Save the reverse route.
+  auto fromToEntry = _visibilityInfoWithRelays[std::make_pair(_from, _to)];
+  auto toFromPath = fromToEntry.route;
+  std::reverse(std::begin(toFromPath), std::end(toFromPath));
+
   _visibilityInfoWithRelays[std::make_pair(_to, _from)] =
-    std::min(srcToDstCost, bestRouteCost);
+    {fromToEntry.cost, toFromPath, fromToEntry.posLastBreadcrumb,
+     fromToEntry.posFirstBreadcrumb, fromToEntry.greatestDistanceSingleHop};
 
   return true;
 }
 
+//////////////////////////////////////////////////
+double VisibilityTable::MaxDistanceSingleHop(
+  const VisibilityInfo &_visibilityInfoWithRelays,
+  const std::vector<ignition::math::graph::VertexId> &_relaySequence,
+  const ignition::math::graph::VertexId &_to) const
+{
+  if (_relaySequence.size() < 2)
+    return 0;
+
+  auto firstBreadcrumb = _relaySequence.at(0);
+  auto secondBreadcrumb = _relaySequence.at(1);
+
+  // We only consider the first breadcrumb stored in the tile.
```

Comment:
Is it accurate that additional breadcrumbs in a tile have no bearing on the position max distance single hop?

Reply:
> Is it accurate that additional breadcrumbs in a tile have no bearing on a max distance single hop from that tile?
> 
> e.g.
> Placing a breadcrumb at the beginning and end of a long tile would have an adverse effect on communication beyond the max hop distance would be unable to factor in the breadcrumb at the end of the long tile.

That's correct, we're not taking into account additional breadcrumbs within the same tile.

Reply:
This could be related to https://github.com/osrf/subt/issues/587